src/telemetry/logs.ts

```diff
@@ -1,105 +1,243 @@
+type ContextualMetadata = {
+  includeContextMetadata?: boolean;
+  workflowUUID: string;
+  authenticatedUser: string;
+  traceId: string;
+  spanId: string;
+};
+
+interface StackTrace {
+  stack?: string;
 }
 
-/* This is a wrapper around a global Winston logger. It holds a reference to the global logger.
- * This class is expected to be instantiated by new DBOSContext such that they can share context information.
- **/
-export class Logger {
-  // Eventually this object will implement one of our TelemetrySignal interface
-  readonly metadata: object = {};
-  constructor(private readonly globalLogger: WinstonLogger, private readonly ctx: DBOSContext) {
-    if (this.globalLogger.addContextMetadata) {
-      this.metadata = {
-        workflowUUID: this.ctx.workflowUUID,
-        authenticatedUser: this.ctx.authenticatedUser,
-        traceId: this.ctx.span.spanContext().traceId,
-        spanId: this.ctx.span.spanContext().spanId,
-      };
+// Wrap around the winston logger to support configuration and access to our telemetry collector
+export interface IGlobalLogger extends IWinstonLogger {
+  readonly addContextMetadata: boolean;
+  readonly logger: IWinstonLogger;
+  readonly telemetryCollector: TelemetryCollector;
+}
+
+export class GlobalLogger {
+  private readonly logger: IWinstonLogger;
+  readonly addContextMetadata: boolean;
+
+  constructor(
+    private readonly telemetryCollector?: TelemetryCollector,
+    config?: LoggerConfig
+  ) {
+    const winstonTransports: TransportStream[] = [];
+    winstonTransports.push(
+      new transports.Console({
+        format: consoleFormat,
+        level: config?.logLevel || "info",
+        silent: config?.silent || false,
+      })
+    );
+    // Only enable the OTLP transport if we have a telemetry collector and an exporter
+    if (this.telemetryCollector?.exporter) {
+      winstonTransports.push(new OTLPLogQueueTransport(this.telemetryCollector));
     }
+    this.logger = createLogger({ transports: winstonTransports });
+    this.addContextMetadata = config?.addContextMetadata || false;
   }
 
-  info(message: string): void {
-    this.globalLogger.info(message, this.metadata);
+  // We use this form of winston logging methods: `(message: string, ...meta: any[])`. See node_modules/winston/index.d.ts
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  info(logEntry: any, metadata?: ContextualMetadata): void {
+    if (typeof logEntry === "string") {
+      this.logger.info(logEntry, metadata);
+    } else {
+      this.logger.info(JSON.stringify(logEntry), metadata);
+    }
   }
 
-  debug(message: string): void {
-    this.globalLogger.debug(message, this.metadata);
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  debug(logEntry: any, metadata?: ContextualMetadata): void {
+    if (typeof logEntry === "string") {
+      this.logger.debug(logEntry, metadata);
+    } else {
+      this.logger.debug(JSON.stringify(logEntry), metadata);
```

Comment:
I suppose we wouldn't catch errors thrown from stringify (for circular references, etc.)?

Reply:
The contract would be that we ask users to give use objects that can be `JSON.stringify`-ed. Under the hood, Winston seems to be more flexible, but it is also really complicated..